Objects/longobject.c

```diff
@@ -3611,24 +3617,45 @@ long_richcompare(PyObject *self, PyObject *other, int op)
     Py_RETURN_RICHCOMPARE(result, 0, op);
 }
 
+void
+_PyLong_ExactDealloc(PyObject *self)
+{
+    if (_PyLong_IsCompact((PyLongObject *)self)) {
+        _Py_FREELIST_FREE(ints, self, PyObject_Free);
+        return;
+    }
+    PyObject_Free(self);
+}
+
 static void
 long_dealloc(PyObject *self)
 {
-    /* This should never get called, but we also don't want to SEGV if
-     * we accidentally decref small Ints out of existence. Instead,
-     * since small Ints are immortal, re-set the reference count.
-     */
     PyLongObject *pylong = (PyLongObject*)self;
+
     if (pylong && _PyLong_IsCompact(pylong)) {
         stwodigits ival = medium_value(pylong);
         if (IS_SMALL_INT(ival)) {
             PyLongObject *small_pylong = (PyLongObject *)get_small_int((sdigit)ival);
             if (pylong == small_pylong) {
+                /* This should never get called, but we also don't want to SEGV if
+                * we accidentally decref small Ints out of existence. Instead,
+                * since small Ints are immortal, re-set the reference count.
+                */
+                // can we remove the next two lines? the immortal objects now have a fixed refcount
+                // in particular in the free-threading build this seeems safe
                 _Py_SetImmortal(self);
                 return;
             }
         }
     }
+
+    if (PyLong_CheckExact(self)) {
+        if (_PyLong_IsCompact((PyLongObject *)self)) {
```

Comment:
@Fidget-Spinner The `_PyLong_IsCompact` check has already been done in this method, can we move this into the `if (pylong && _PyLong_IsCompact(pylong))` part?

Also, can we remove the `pylong &&` part? I think `pylong` can never be NULL. (`PyLong_CheckExact` assumes the pointer is not NULL I think)